File: src/Parser.py
import Tokens as Token
from Parser_types import *
class Parser:
    def __init__(self, tokens=None):
        self.tokens = tokens
        self.pos = 0

    def error(self, message="Invalid syntax"):
        raise Exception(message)

    def current_token(self):
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def eat(self, token_type):
        if self.current_token() and self.current_token().type == token_type:
            self.pos += 1
        else:
            self.error(f"Expected token: {token_type}, found: {self.current_token()}")
    def get_statement(self):
        if self.current_token().type == Token.TOKENTYPE.COMMENT:
            self.eat(Token.TOKENTYPE.COMMENT)
        elif self.current_token().type == Token.TOKENTYPE.WHILE:
            return self.parse_while_statement()
        elif self.current_token().type == Token.TOKENTYPE.FOR:
            return self.parse_for_loop()
        
        elif self.current_token().type == Token.TOKENTYPE.FUNCTION_DECLARATION:
            return self.parse_function_declaration()
        elif self.current_token().type == Token.TOKENTYPE.VAR:
            return self.parse_variable_declaration()
        elif self.current_token().type in Token.AUGMENTED_ASSIGNMENT_OPERATORS:
            return self.parse_augmented_assignment()
        elif self.current_token().type == Token.TOKENTYPE.NAME:
            lookahead_token = self.tokens[self.pos + 1] if self.pos + 1 < len(self.tokens) else None
            if lookahead_token and lookahead_token.type in Token.AUGMENTED_ASSIGNMENT_OPERATORS:
                return self.parse_augmented_assignment()
            elif lookahead_token and lookahead_token.type == Token.TOKENTYPE.LPAREN:
                return self.parse_function_call()
            elif lookahead_token and lookahead_token.type == Token.TOKENTYPE.DOT:
                return self.parse_method_call(Variable(self.current_token().value))
            else:
                return self.parse_assignment()
        elif self.current_token().type == Token.TOKENTYPE.IF:
            return self.parse_if_statement()
        elif self.current_token().type == Token.TOKENTYPE.ELSE:
            return self.parse_else_statement()
        else:
            self.error("Unexpected token " + self.current_token().type)
    def parse(self):
        statements = []
        while self.current_token() is not None and self.current_token().type != Token.TOKENTYPE.EOF:
            statements.append(self.get_statement())
        return statements

    def parse_for_loop(self):
        self.eat(Token.TOKENTYPE.FOR)
        self.eat(Token.TOKENTYPE.LPAREN)

        # Initial expression - can be a variable declaration or any expression
        init = None
        if self.current_token().type == Token.TOKENTYPE.VAR:
            init = self.parse_variable_declaration()
        else:
            init = self.parse_expression()
            if isinstance(init, Expression):
                init = init.expr  # unwrap expression from statement if needed
        self.eat(Token.TOKENTYPE.SEMICOLON)

        # Condition - any expression
        condition = self.parse_expression()
        self.eat(Token.TOKENTYPE.SEMICOLON)

        # Update - can be an assignment or any expression
        update = None
        if self.current_token().type == Token.TOKENTYPE.NAME:
            if self.tokens[self.pos + 1].type in Token.AUGMENTED_ASSIGNMENT_OPERATORS:
                update = self.parse_augmented_assignment()
            else:
                update = self.parse_assignment()
        else:
            update = self.parse_expression()

        self.eat(Token.TOKENTYPE.RPAREN)
        self.eat(Token.TOKENTYPE.LBRACE)
        body = self.parse_block()
        self.eat(Token.TOKENTYPE.RBRACE)

        return ForStatement(init, condition, update, body)

    def parse_function_declaration(self):
        self.eat(Token.TOKENTYPE.FUNCTION_DECLARATION)
        func_name = self.current_token().value
        self.eat(Token.TOKENTYPE.NAME)
        self.eat(Token.TOKENTYPE.LPAREN)
        parameters = self.parse_parameters()
        self.eat(Token.TOKENTYPE.RPAREN)
        self.eat(Token.TOKENTYPE.LBRACE)
        body = self.parse_block()
        self.eat(Token.TOKENTYPE.RBRACE)
        return FunctionDeclaration(func_name, parameters, body)

    def parse_function_call(self):
        func_name = self.current_token().value
        self.eat(Token.TOKENTYPE.NAME)
        self.eat(Token.TOKENTYPE.LPAREN)
        arguments = self.parse_arguments()
        self.eat(Token.TOKENTYPE.RPAREN)
        return FunctionCall(func_name, arguments)

    def parse_arguments(self):
        arguments = []
        while self.current_token().type != Token.TOKENTYPE.RPAREN:
            arguments.append(self.parse_expression())
            if self.current_token().type == Token.TOKENTYPE.COMMA:
                self.eat(Token.TOKENTYPE.COMMA)
        return arguments
    
    def parse_list(self):
        values = []
        while self.current_token().type != Token.TOKENTYPE.RBRACK:
            values.append(self.parse_expression())
            if self.current_token().type == Token.TOKENTYPE.COMMA:
                self.eat(Token.TOKENTYPE.COMMA)
    def parse_parameters(self):
        parameters = []
        while self.current_token().type != Token.TOKENTYPE.RPAREN:
            parameters.append(self.current_token().value)
            self.eat(Token.TOKENTYPE.NAME)
            if self.current_token().type == Token.TOKENTYPE.COMMA:
                self.eat(Token.TOKENTYPE.COMMA)
        return parameters

    def parse_block(self):
        statements = []
        while self.current_token().type != Token.TOKENTYPE.RBRACE:
            if self.current_token().type == Token.TOKENTYPE.COMMENT:
                self.eat(Token.TOKENTYPE.COMMENT)
            elif self.current_token().type == Token.TOKENTYPE.WHILE:
                statements.append(self.parse_while_statement())
            elif self.current_token().type == Token.TOKENTYPE.IF:
                statements.append(self.parse_if_statement())
            
            elif self.current_token().type == Token.TOKENTYPE.RETURN:
                statements.append(self.parse_return_statement())
            elif self.current_token().type == Token.TOKENTYPE.VAR:
                statements.append(self.parse_variable_declaration())
            elif self.current_token().type == Token.TOKENTYPE.NAME:
                lookahead_token = self.tokens[self.pos + 1] if self.pos + 1 < len(self.tokens) else None
                if lookahead_token and lookahead_token.type == Token.TOKENTYPE.LPAREN:
                    statements.append(self.parse_function_call())
                elif lookahead_token and lookahead_token.type in Token.AUGMENTED_ASSIGNMENT_OPERATORS:
                    statements.append(self.parse_augmented_assignment())
                else:
                    statements.append(self.parse_assignment())
            else:
                self.error("Unexpected token in block " + self.current_token().type)
        return statements
    def parse_return_statement(self):
        self.eat(Token.TOKENTYPE.RETURN)
        expr = self.parse_expression()
        return ReturnStatement(expr)

    def parse_variable_declaration(self):
        self.eat(Token.TOKENTYPE.VAR)
        var_name = self.current_token().value
        self.eat(Token.TOKENTYPE.NAME)
        self.eat(Token.TOKENTYPE.EQUALS)
        expr = self.parse_expression()
        return VariableDeclaration(var_name, expr)

    def parse_augmented_assignment(self):
        # get asssignment as variable_name, operator, right hand side
        var_name = self.current_token().value
        self.eat(Token.TOKENTYPE.NAME)
        operator = self.current_token().type
        self.eat(operator)
        expr = self.parse_expression()
        print(var_name,operator,expr)
        return AugmentedAssignment(var_name, operator, expr)
    def parse_assignment(self):
        var_name = self.current_token().value
        self.eat(Token.TOKENTYPE.NAME)
        self.eat(Token.TOKENTYPE.EQUALS)
        expr = self.parse_expression()
        return Assignment(var_name, expr)

    def parse_expression(self):
        node = self.parse_term()
    
        while self.current_token() is not None and self.current_token().type in Token.COMPARISON_OPERATORS:
            token_type = self.current_token().type
            self.eat(token_type)
            node = ComparisonOperation(node, token_type, self.parse_term())

        return node

    def parse_term(self):
        node = self.parse_factor()

        while self.current_token() is not None and self.current_token().type in (Token.TOKENTYPE.PLUS, Token.TOKENTYPE.MINUS, Token.TOKENTYPE.MODULO, Token.TOKENTYPE.CARAT):
            token_type = self.current_token().type
            self.eat(token_type)
            node = BinaryOperation(node, token_type, self.parse_factor())

        return node

    def parse_factor(self):
        node = self.parse_atom()

        while self.current_token() is not None and self.current_token().type in (Token.TOKENTYPE.MUL, Token.TOKENTYPE.DIV):
            token_type = self.current_token().type
            self.eat(token_type)
            node = BinaryOperation(node, token_type, self.parse_atom())

        return node

    def parse_array_literal(self):
        self.eat(Token.TOKENTYPE.LBRACK)
        elements = []
        if self.current_token().type != Token.TOKENTYPE.RBRACK:
            elements.append(self.parse_expression())
            while self.current_token().type == Token.TOKENTYPE.COMMA:
                self.eat(Token.TOKENTYPE.COMMA)
                elements.append(self.parse_expression())
        self.eat(Token.TOKENTYPE.RBRACK)
        return ArrayLiteral(elements)
    def parse_method_call(self, variable):
        self.eat(Token.TOKENTYPE.NAME)
        self.eat(Token.TOKENTYPE.DOT)
        method_name = self.current_token().value
        self.eat(Token.TOKENTYPE.NAME)
        self.eat(Token.TOKENTYPE.LPAREN)
        arguments = self.parse_arguments()
        self.eat(Token.TOKENTYPE.RPAREN)
        return MethodCall(variable, method_name, arguments)
    def parse_atom(self):
        token = self.current_token()
        if token.type == Token.TOKENTYPE.BANG:
            self.eat(Token.TOKENTYPE.BANG)
            operand = self.parse_atom()
            return UnaryOperation(Token.TOKENTYPE.BANG,operand)
        elif token.type == Token.TOKENTYPE.MINUS:
            self.eat(Token.TOKENTYPE.MINUS)
            operand = self.parse_atom()
            #print(UnaryOperation(operand, Token.TOKENTYPE.MINUS))
            return UnaryOperation(Token.TOKENTYPE.MINUS, operand)
        elif token.type == Token.TOKENTYPE.INTEGER:
            self.eat(Token.TOKENTYPE.INTEGER)
            return NumberLiteral("int", token.value)
        elif token.type == Token.TOKENTYPE.STRING:
            self.eat(Token.TOKENTYPE.STRING)
            token.value = token.value[1:-1]
            return StringLiteral(token.value)
        elif token.type == Token.TOKENTYPE.FLOAT:
            self.eat(Token.TOKENTYPE.FLOAT)
            return NumberLiteral("float", token.value)
        elif token.type == Token.TOKENTYPE.TRUE:
            self.eat(Token.TOKENTYPE.TRUE)
            return BooleanLiteral("true")
        elif token.type == Token.TOKENTYPE.FALSE:
            self.eat(Token.TOKENTYPE.FALSE)
            return BooleanLiteral("false")
        elif token.type == Token.TOKENTYPE.LBRACK:
            return self.parse_array_literal()
        elif token.type == Token.TOKENTYPE.NAME:
            # Check if it's a function call
            lookahead_token = self.tokens[self.pos + 1] if self.pos + 1 < len(self.tokens) else None
            if lookahead_token and lookahead_token.type == Token.TOKENTYPE.LPAREN:
                return self.parse_function_call()
            elif lookahead_token and lookahead_token.type == Token.TOKENTYPE.DOT:
                return self.parse_method_call(Variable(token.value))
            else:
                self.eat(Token.TOKENTYPE.NAME)
                return Variable(token.value)
        elif token.type == Token.TOKENTYPE.LPAREN:
            self.eat(Token.TOKENTYPE.LPAREN)
            node = self.parse_expression()
            self.eat(Token.TOKENTYPE.RPAREN)
            return node
        else:
            self.error(f"Unknown atom {token}")
    def parse_if_statement(self):
        self.eat(Token.TOKENTYPE.IF)
        self.eat(Token.TOKENTYPE.LPAREN)
        condition = self.parse_expression()
        self.eat(Token.TOKENTYPE.RPAREN)
        self.eat(Token.TOKENTYPE.LBRACE)
        if_body = self.parse_block()
        self.eat(Token.TOKENTYPE.RBRACE)
        else_body = None
        if self.current_token() and self.current_token().type == Token.TOKENTYPE.ELSE:
            else_body = self.parse_else_statement()
        return IfStatement(condition, if_body, else_body)
    def parse_else_statement(self):
        self.eat(Token.TOKENTYPE.ELSE)
        self.eat(Token.TOKENTYPE.LBRACE)
        else_body = self.parse_block()
        self.eat(Token.TOKENTYPE.RBRACE)
        return else_body
    def parse_while_statement(self):
        self.eat(Token.TOKENTYPE.WHILE)
        self.eat(Token.TOKENTYPE.LPAREN)
        condition = self.parse_expression()
        self.eat(Token.TOKENTYPE.RPAREN)
        self.eat(Token.TOKENTYPE.LBRACE)
        body = self.parse_block()
        self.eat(Token.TOKENTYPE.RBRACE)
        return WhileStatement(condition, body)

def parse_program(tokens):
    parser = Parser(tokens)
    return parser.parse()

def simple_ast_format(ast):
    # This function is used to format the AST for printing

    def format_node(node, indent=0):
        if isinstance(node, list):
            return "".join([format_node(item, indent) for item in node])
        elif isinstance(node, dict):
            string = ""
            for key, value in node.items():
                string += "\t" * indent + str(key) + "\n"
                string += format_node(value, indent + 1)
            return string
        else:
            return "\t" * indent + str(node) + "\n"
    return format_node(ast)


def parse_program(tokens):
    parser = Parser(tokens)
    return parser.parse()


if __name__ == "__main__":
    import Lexer, json
    program =  """
    var x = [1,2,3,4,5,6,7,8,9,10]
    print(x.sort())

    """
    tokens = Lexer.tokenize(program)
    print(tokens)
    print(simple_ast_format(parse_program(tokens)))

File: src/cpp_version/Parser.cpp


#include "ParserClass.h"
#include "Lexer.h"

// Constructor implementation
Parser::Parser(const std::vector<Token>& tokens) : tokens(tokens), pos(0) {}

// Implementations of other methods
void Parser::error(const std::string& message) {
    throw std::runtime_error(message);
}

Token Parser::current_token() {
    if (pos < tokens.size()) {
        return tokens[pos];
    }
    return Token(TOKENTYPE::EOF_TOKEN, "");
}

void Parser::eat(TOKENTYPE token_type) {
    if (current_token().type == token_type) {
        pos++;
    } else {
        error("Expected token: " + TOKENTYPE_to_string(token_type) + ", found: " + TOKENTYPE_to_string(current_token().type) + " at: " + std::to_string(pos));
    }
}

Statement* Parser::get_statement() {
    if (current_token().type == TOKENTYPE::VAR) {
        return parse_variable_declaration();
    } else if (current_token().type == TOKENTYPE::RETURN) {
        return parse_return_statement();
    } else if (current_token().type == TOKENTYPE::IF) {
        return parse_if_statement();
    } else if (current_token().type == TOKENTYPE::WHILE) {
        return parse_while_statement();
    } else if (current_token().type == TOKENTYPE::FUNCTION_DECLARATION) {
        return parse_function_declaration();
    } else {
        return parse_expression_statement();
    }
}

Statement* Parser::parse_variable_declaration() {
    eat(TOKENTYPE::VAR);
    std::string variable_name = current_token().value;
    eat(TOKENTYPE::NAME);
    eat(TOKENTYPE::EQUALS);
    Expression* expr = parse_expression();
    return new VariableDeclaration(variable_name, expr);
}

Statement* Parser::parse_return_statement() {
    eat(TOKENTYPE::RETURN);
    Expression* expr = parse_expression();
    return new ReturnStatement(expr);
}

Statement* Parser::parse_expression_statement() {
    Expression* expr = parse_expression();
    return new ExpressionStatement(expr);
}

std::vector<std::string> Parser::parse_parameters() {
    std::vector<std::string> parameters;
    eat(TOKENTYPE::LPAREN);
    if (current_token().type == TOKENTYPE::RPAREN) {
        eat(TOKENTYPE::RPAREN);
        return parameters;
    }
    parameters.push_back(current_token().value);
    eat(TOKENTYPE::NAME);
    while (current_token().type == TOKENTYPE::COMMA) {
        eat(TOKENTYPE::COMMA);
        parameters.push_back(current_token().value);
        eat(TOKENTYPE::NAME);
    }
    eat(TOKENTYPE::RPAREN);
    return parameters;
}

Block* Parser::parse_block() {
    eat(TOKENTYPE::LBRACE);
    std::vector<Statement*> statements;
    while (current_token().type != TOKENTYPE::RBRACE) {
        statements.push_back(get_statement());
    }
    eat(TOKENTYPE::RBRACE);
    return new Block(statements);
}

Statement* Parser::parse_if_statement() {
    eat(TOKENTYPE::IF);
    Expression* condition = parse_expression();
    Block* if_block = parse_block();
    Block* else_block = nullptr;
    if (current_token().type == TOKENTYPE::ELSE) {
        else_block = parse_else_statement();
    }
    return new IfStatement(condition, if_block, else_block);
}

Block* Parser::parse_else_statement() {
    eat(TOKENTYPE::ELSE);
    return parse_block();
}

Expression* Parser::parse_atom(){

    Token token = current_token();
    if (token.type == TOKENTYPE::INTEGER) {
        eat(TOKENTYPE::INTEGER);
        return new NumberLiteral("int", token.value);
    } else if (token.type == TOKENTYPE::FLOAT) {
        eat(TOKENTYPE::FLOAT);
        return new NumberLiteral("float", token.value);
    } else if (token.type == TOKENTYPE::STRING) {
        eat(TOKENTYPE::STRING);
        return new StringLiteral(token.value);
    } else if (token.type == TOKENTYPE::TRUE) {
        eat(TOKENTYPE::TRUE);
        return new BooleanLiteral(true);
    } else if (token.type == TOKENTYPE::FALSE) {
        eat(TOKENTYPE::FALSE);
        return new BooleanLiteral(false);
    } else if (token.type == TOKENTYPE::NAME) {
        eat(TOKENTYPE::NAME);
        return new Variable(token.value);
    } else if (token.type == TOKENTYPE::LPAREN) {
        eat(TOKENTYPE::LPAREN);
        Expression* expr = parse_expression();
        eat(TOKENTYPE::RPAREN);
        return expr;
    }
}   

Expression* Parser::parse_factor(){
    /*Python example
    def parse_factor(self):
        node = self.parse_atom()

        while self.current_token() is not None and self.current_token().type in (Token.TOKENTYPE.MUL, Token.TOKENTYPE.DIV):
            token_type = self.current_token().type
            self.eat(token_type)
            node = BinaryOperation(node, token_type, self.parse_atom())

        return node
    */

    Expression* expr = parse_atom();
    while (current_token().type == TOKENTYPE::MUL || current_token().type == TOKENTYPE::DIV) {
        if (current_token().type == TOKENTYPE::MUL) {
            eat(TOKENTYPE::MUL);
            expr = new BinaryOperation(expr, TOKENTYPE::MUL, parse_atom());
        } else if (current_token().type == TOKENTYPE::DIV) {
            eat(TOKENTYPE::DIV);
            expr = new BinaryOperation(expr, TOKENTYPE::DIV , parse_atom());

        }
    }
    return expr;
}

Statement* Parser::parse_while_statement() {
    eat(TOKENTYPE::WHILE);
    Expression* condition = parse_expression();
    Block* while_block = parse_block();
    return new WhileStatement(condition, while_block);
}

FunctionDeclaration* Parser::parse_function_declaration() {
    eat(TOKENTYPE::FUNCTION_DECLARATION);
    std::string name = current_token().value;
    eat(TOKENTYPE::NAME);
    std::vector<std::string> parameters = parse_parameters();
    Block* body = parse_block();
    return new FunctionDeclaration(name, parameters, body);
}

FunctionCall* Parser::parse_function_call() {
    std::string name = current_token().value;
    eat(TOKENTYPE::NAME);
    eat(TOKENTYPE::LPAREN);
    std::vector<Expression*> arguments = parse_arguments();
    eat(TOKENTYPE::RPAREN);
    return new FunctionCall(name, arguments);
}

std::vector<Expression*> Parser::parse_arguments() {
    std::vector<Expression*> arguments;
    while (current_token().type != TOKENTYPE::RPAREN) {
        std::cout << arguments.size() << std::endl;
        arguments.push_back(parse_expression());
        if (current_token().type == TOKENTYPE::COMMA) {
            eat(TOKENTYPE::COMMA);
        }
        else if (current_token().type == TOKENTYPE::RPAREN) {
            break;
        }
    }
    std::cout << arguments.size() << std::endl;
    return arguments;
}


Expression* Parser::parse_expression() {
    Expression* expr = parse_term();
    while (current_token().type == TOKENTYPE::IS_EQUAL ||
           current_token().type == TOKENTYPE::LESS_THAN ||
           current_token().type == TOKENTYPE::GREATER_THAN ||
           current_token().type == TOKENTYPE::LESS_THAN_OR_EQUAL ||
           current_token().type == TOKENTYPE::GREATER_THAN_OR_EQUAL ||
           current_token().type == TOKENTYPE::NOT_EQUAL) {
        TOKENTYPE token_type = current_token().type;
        eat(token_type);
        expr = new ComparisonOperation(expr, token_type, parse_term());
    }
    return expr;
}

Expression* Parser::parse_term() {
    Expression* expr = parse_factor();
    while (current_token().type == TOKENTYPE::MUL || current_token().type == TOKENTYPE::DIV) {
        if (current_token().type == TOKENTYPE::MUL) {
            eat(TOKENTYPE::MUL);
            expr = new BinaryOperation(expr, TOKENTYPE::MUL, parse_factor());
        } else if (current_token().type == TOKENTYPE::DIV) {
            eat(TOKENTYPE::DIV);
            expr = new BinaryOperation(expr, TOKENTYPE::DIV , parse_factor());
        }
        
    }
    return expr;
}

std::vector<Statement*> Parser::parse() {
    std::vector<Statement*> statements;
    while (current_token().type != TOKENTYPE::EOF_TOKEN) {
        statements.push_back(get_statement());
    }
    return statements;
}

int main(){
    std::string program = R"(
        var a = 5
        mkfunc add(a, b) {
            return a + b
        }
        var b = add(5, 10)
        print(b)
    )";
    Lexer lexer(program);
    std::vector<Token> tokens = lexer.tokenize();
    for (auto token : tokens) {
        std::cout << token.toString() << std::endl;
    }

    Parser parser(tokens);
    std::vector<Statement*> statements = parser.parse();
    for (auto statement : statements) {
        std::cout << statement->toString() << std::endl;
    }
}
